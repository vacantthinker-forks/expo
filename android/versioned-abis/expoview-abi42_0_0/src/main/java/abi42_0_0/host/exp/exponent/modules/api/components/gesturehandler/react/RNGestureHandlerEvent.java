package abi42_0_0.host.exp.exponent.modules.api.components.gesturehandler.react;

import androidx.core.util.Pools;

import abi42_0_0.com.facebook.react.bridge.Arguments;
import abi42_0_0.com.facebook.react.bridge.WritableMap;
import abi42_0_0.com.facebook.react.uimanager.events.Event;
import abi42_0_0.com.facebook.react.uimanager.events.RCTEventEmitter;
import abi42_0_0.host.exp.exponent.modules.api.components.gesturehandler.GestureHandler;

import androidx.annotation.Nullable;

public class RNGestureHandlerEvent extends Event<RNGestureHandlerEvent> {

  public static final String EVENT_NAME = "onGestureHandlerEvent";

  private static final int TOUCH_EVENTS_POOL_SIZE = 7; // magic

  private static final Pools.SynchronizedPool<RNGestureHandlerEvent> EVENTS_POOL =
          new Pools.SynchronizedPool<>(TOUCH_EVENTS_POOL_SIZE);

  public static RNGestureHandlerEvent obtain(
          GestureHandler handler,
          @Nullable RNGestureHandlerEventDataExtractor dataExtractor) {
    RNGestureHandlerEvent event = EVENTS_POOL.acquire();
    if (event == null) {
      event = new RNGestureHandlerEvent();
    }
    event.init(handler, dataExtractor);
    return event;
  }

  private WritableMap mExtraData;
  private short mCoalescingKey;

  private RNGestureHandlerEvent() {
  }

  private void init(
          GestureHandler handler,
          @Nullable RNGestureHandlerEventDataExtractor dataExtractor) {
    super.init(handler.getView().getId());
    mExtraData = Arguments.createMap();
    if (dataExtractor != null) {
      dataExtractor.extractEventData(handler, mExtraData);
    }
    mExtraData.putInt("handlerTag", handler.getTag());
    mExtraData.putInt("state", handler.getState());
    mCoalescingKey = handler.getEventCoalescingKey();
  }

  @Override
  public void onDispose() {
    mExtraData = null;
    EVENTS_POOL.release(this);
  }

  @Override
  public String getEventName() {
    return EVENT_NAME;
  }

  @Override
  public boolean canCoalesce() {
    return true;
  }

  @Override
  public short getCoalescingKey() {
    return mCoalescingKey;
  }

  @Override
  public void dispatch(RCTEventEmitter rctEventEmitter) {
    rctEventEmitter.receiveEvent(getViewTag(), EVENT_NAME, mExtraData);
  }
}
