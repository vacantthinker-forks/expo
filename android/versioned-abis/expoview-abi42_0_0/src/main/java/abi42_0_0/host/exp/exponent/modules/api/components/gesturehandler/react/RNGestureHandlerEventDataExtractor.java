package abi42_0_0.host.exp.exponent.modules.api.components.gesturehandler.react;

import abi42_0_0.com.facebook.react.bridge.WritableMap;
import abi42_0_0.host.exp.exponent.modules.api.components.gesturehandler.GestureHandler;

public interface RNGestureHandlerEventDataExtractor<T extends GestureHandler> {
  void extractEventData(T handler, WritableMap eventData);
}
