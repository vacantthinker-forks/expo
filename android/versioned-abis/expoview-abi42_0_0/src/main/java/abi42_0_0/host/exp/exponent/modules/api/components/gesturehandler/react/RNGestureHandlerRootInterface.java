package abi42_0_0.host.exp.exponent.modules.api.components.gesturehandler.react;

import androidx.annotation.Nullable;

public interface RNGestureHandlerRootInterface {
  @Nullable RNGestureHandlerRootHelper getRootHelper();
}
