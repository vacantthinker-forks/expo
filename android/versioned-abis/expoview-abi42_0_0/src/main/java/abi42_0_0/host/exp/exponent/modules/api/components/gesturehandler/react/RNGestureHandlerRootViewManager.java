package abi42_0_0.host.exp.exponent.modules.api.components.gesturehandler.react;

import abi42_0_0.com.facebook.react.common.MapBuilder;
import abi42_0_0.com.facebook.react.module.annotations.ReactModule;
import abi42_0_0.com.facebook.react.uimanager.ThemedReactContext;
import abi42_0_0.com.facebook.react.uimanager.ViewGroupManager;

import java.util.Map;

import androidx.annotation.Nullable;

/**
 * React native's view manager used for creating instances of {@link }RNGestureHandlerRootView}. It
 * is being used by projects using react-native-navigation where for each screen new root view need
 * to be provided.
 */
@ReactModule(name = RNGestureHandlerRootViewManager.REACT_CLASS)
public class RNGestureHandlerRootViewManager extends ViewGroupManager<RNGestureHandlerRootView> {

  public static final String REACT_CLASS = "GestureHandlerRootView";

  @Override
  public String getName() {
    return REACT_CLASS;
  }

  @Override
  protected RNGestureHandlerRootView createViewInstance(ThemedReactContext reactContext) {
    return new RNGestureHandlerRootView(reactContext);
  }

  @Override
  public void onDropViewInstance(RNGestureHandlerRootView view) {
    view.tearDown();
  }

  /**
   * The following event configuration is necessary even if you are not using
   * GestureHandlerRootView component directly.
   */
  @Override
  public @Nullable Map getExportedCustomDirectEventTypeConstants() {
    return MapBuilder.of(
            RNGestureHandlerEvent.EVENT_NAME,
            MapBuilder.of("registrationName", RNGestureHandlerEvent.EVENT_NAME),
            RNGestureHandlerStateChangeEvent.EVENT_NAME,
            MapBuilder.of("registrationName", RNGestureHandlerStateChangeEvent.EVENT_NAME));
  }
}
