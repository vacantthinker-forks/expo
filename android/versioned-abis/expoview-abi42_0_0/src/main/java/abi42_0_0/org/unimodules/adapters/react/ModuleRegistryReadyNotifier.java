package abi42_0_0.org.unimodules.adapters.react;

import abi42_0_0.com.facebook.react.bridge.BaseJavaModule;
import abi42_0_0.com.facebook.react.bridge.NativeModule;

import abi42_0_0.org.unimodules.core.ModuleRegistry;

/**
 * {@link ModuleRegistryReadyNotifier} is exported as a native module
 * to React Native and when {@link abi42_0_0.com.facebook.react.ReactInstanceManager}
 * notifies {@link abi42_0_0.com.facebook.react.bridge.NativeModule} of being ready
 * ({@link NativeModule#initialize()}) it delegates the call to {@link ModuleRegistry}.
 */
public class ModuleRegistryReadyNotifier extends BaseJavaModule {
  private ModuleRegistry mModuleRegistry;

  public ModuleRegistryReadyNotifier(ModuleRegistry moduleRegistry) {
    mModuleRegistry = moduleRegistry;
  }

  @Override
  public String getName() {
    return null;
  }

  @Override
  public void initialize() {
    mModuleRegistry.ensureIsInitialized();
  }
}
