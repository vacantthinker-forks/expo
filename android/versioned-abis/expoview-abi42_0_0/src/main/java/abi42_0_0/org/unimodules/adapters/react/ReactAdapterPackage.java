package abi42_0_0.org.unimodules.adapters.react;

import android.content.Context;

import abi42_0_0.com.facebook.react.bridge.ReactContext;

import abi42_0_0.org.unimodules.adapters.react.permissions.PermissionsService;
import abi42_0_0.org.unimodules.adapters.react.services.CookieManagerModule;
import abi42_0_0.org.unimodules.adapters.react.services.EventEmitterModule;
import abi42_0_0.org.unimodules.adapters.react.services.FontManagerModule;
import abi42_0_0.org.unimodules.adapters.react.services.RuntimeEnvironmentModule;
import abi42_0_0.org.unimodules.adapters.react.services.UIManagerModuleWrapper;
import abi42_0_0.org.unimodules.core.BasePackage;
import abi42_0_0.org.unimodules.core.interfaces.InternalModule;
import abi42_0_0.org.unimodules.core.interfaces.Package;

import java.util.Arrays;
import java.util.List;

/**
 * A {@link Package} creating modules provided with the @unimodules/react-native-adapter package.
 */
public class ReactAdapterPackage extends BasePackage {

  @Override
  public List<InternalModule> createInternalModules(Context context) {
    // We can force-cast here, because this package will only be used in React Native context.
    ReactContext reactContext = (ReactContext) context;
    return Arrays.asList(
        new CookieManagerModule(reactContext),
        new UIManagerModuleWrapper(reactContext),
        new EventEmitterModule(reactContext),
        new FontManagerModule(),
        new RuntimeEnvironmentModule(),
        new PermissionsService(reactContext)
    );
  }
}
