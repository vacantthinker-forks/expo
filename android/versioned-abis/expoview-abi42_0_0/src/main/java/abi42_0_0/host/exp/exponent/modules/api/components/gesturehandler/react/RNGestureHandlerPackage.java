package abi42_0_0.host.exp.exponent.modules.api.components.gesturehandler.react;

import abi42_0_0.com.facebook.react.ReactPackage;
import abi42_0_0.com.facebook.react.bridge.NativeModule;
import abi42_0_0.com.facebook.react.bridge.ReactApplicationContext;
import abi42_0_0.com.facebook.react.common.MapBuilder;
import abi42_0_0.com.facebook.react.uimanager.ThemedReactContext;
import abi42_0_0.com.facebook.react.uimanager.ViewGroupManager;
import abi42_0_0.com.facebook.react.uimanager.ViewManager;
import abi42_0_0.com.facebook.react.views.view.ReactViewManager;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

import androidx.annotation.Nullable;

public class RNGestureHandlerPackage implements ReactPackage {

  @Override
  public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
    return Arrays.<NativeModule>asList(new RNGestureHandlerModule(reactContext));
  }

  @Override
  public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
    return Arrays.<ViewManager>asList(
            new RNGestureHandlerRootViewManager(),
            new RNGestureHandlerButtonViewManager());
  }
}
