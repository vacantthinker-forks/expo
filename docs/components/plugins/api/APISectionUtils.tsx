import { css } from '@emotion/react';
import { theme } from '@expo/styleguide';
import React from 'react';
import ReactMarkdown from 'react-markdown';

import { Code, InlineCode } from '~/components/base/code';
import { H4 } from '~/components/base/headings';
import Link from '~/components/base/link';
import { LI, UL } from '~/components/base/list';
import { B, P, Quote } from '~/components/base/paragraph';
import {
  CommentData,
  MethodParamData,
  TypeDefinitionData,
} from '~/components/plugins/api/APIDataTypes';

export enum TypeDocKind {
  Enum = 4,
  Variable = 32,
  Function = 64,
  Class = 128,
  Interface = 256,
  Property = 1024,
  TypeAlias = 4194304,
}

export type MDRenderers = React.ComponentProps<typeof ReactMarkdown>['renderers'];

export const mdRenderers: MDRenderers = {
  blockquote: ({ children }) => (
    <Quote>
      {React.Children.map(children, child =>
        child.type.name === 'paragraph' ? child.props.children : child
      )}
    </Quote>
  ),
  code: ({ value, language }) => <Code className={`language-${language}`}>{value}</Code>,
  heading: ({ children }) => <H4>{children}</H4>,
  inlineCode: ({ value }) => <InlineCode>{value}</InlineCode>,
  list: ({ children }) => <UL>{children}</UL>,
  listItem: ({ children }) => <LI>{children}</LI>,
  link: ({ href, children }) => <Link href={href}>{children}</Link>,
  paragraph: ({ children }) => (children ? <P>{children}</P> : null),
  strong: ({ children }) => <B>{children}</B>,
  text: ({ value }) => (value ? <span>{value}</span> : null),
};

export const mdInlineRenderers: MDRenderers = {
  ...mdRenderers,
  paragraph: ({ children }) => (children ? <span>{children}</span> : null),
};

const nonLinkableTypes = [
  'Date',
  'Error',
  'Omit',
  'Promise',
  'React.FC',
  'StyleProp',
  'T',
  'TaskOptions',
  'Uint8Array',
  'ViewStyle',
];

export const resolveTypeName = ({
  elements,
  elementType,
  name,
  type,
  types,
  typeArguments,
  declaration,
  value,
  queryType,
}: TypeDefinitionData): string | JSX.Element | (string | JSX.Element)[] => {
  if (name) {
    if (type === 'reference') {
      if (typeArguments) {
        if (name === 'Record' || name === 'React.ComponentProps') {
          return (
            <>
              {name}&lt;
              {typeArguments.map((type, index) => (
                <span key={`record-type-${index}`}>
                  {resolveTypeName(type)}
                  {index !== typeArguments.length - 1 ? ', ' : null}
                </span>
              ))}
              &gt;
            </>
          );
        } else {
          return (
            <>
              {nonLinkableTypes.includes(name) ? (
                name
              ) : (
                <Link href={`#${name.toLowerCase()}`} key={`type-link-${name}`}>
                  {name}
                </Link>
              )}
              &lt;
              {typeArguments.map((type, index) => (
                <span key={`${name}-nested-type-${index}`}>
                  {resolveTypeName(type)}
                  {index !== typeArguments.length - 1 ? ', ' : null}
                </span>
              ))}
              &gt;
            </>
          );
        }
      } else {
        if (nonLinkableTypes.includes(name)) {
          return name;
        } else {
          return (
            <Link href={`#${name.toLowerCase()}`} key={`type-link-${name}`}>
              {name}
            </Link>
          );
        }
      }
    } else {
      return name;
    }
  } else if (elementType?.name) {
    if (elementType.type === 'reference') {
      if (nonLinkableTypes.includes(elementType.name)) {
        return elementType.name + (type === 'array' && '[]');
      }
      return (
        <Link href={`#${elementType.name?.toLowerCase()}`} key={`type-link-${elementType.name}`}>
          {elementType.name}
          {type === 'array' && '[]'}
        </Link>
      );
    }
    if (type === 'array') {
      return elementType.name + '[]';
    }
    return elementType.name + type;
  } else if (type === 'union' && types?.length) {
    return types.map(resolveTypeName).map((valueToRender, index) => (
      <span key={`union-type-${index}`}>
        {valueToRender}
        {index + 1 !== types.length && ' | '}
      </span>
    ));
  } else if (declaration?.signatures) {
    const baseSignature = declaration.signatures[0];
    if (baseSignature?.parameters?.length) {
      return (
        <>
          (
          {baseSignature.parameters?.map((param, index) => (
            <span key={`param-${index}-${param.name}`}>
              {param.name}: {resolveTypeName(param.type)}
              {index + 1 !== baseSignature.parameters?.length && ', '}
            </span>
          ))}
          ) {'=>'} {resolveTypeName(baseSignature.type)}
        </>
      );
    } else {
      return (
        <>
          {'() =>'} {resolveTypeName(baseSignature.type)}
        </>
      );
    }
  } else if (type === 'tuple' && elements) {
    return (
      <>
        [
        {elements.map((elem, i) => (
          <span key={`tuple-${name}-${i}`}>
            {resolveTypeName(elem)}
            {i + 1 !== elements.length ? ', ' : null}
          </span>
        ))}
        ]
      </>
    );
  } else if (type === 'query' && queryType) {
    return queryType.name;
  } else if (type === 'literal' && value) {
    return `'${value}'`;
  } else if (value === null) {
    return 'null';
  }
  return 'undefined';
};

export const renderParam = ({ comment, name, type }: MethodParamData): JSX.Element => (
  <LI key={`param-${name}`}>
    <B>
      {name} (<InlineCode>{resolveTypeName(type)}</InlineCode>)
    </B>
    <CommentTextBlock comment={comment} renderers={mdInlineRenderers} withDash />
  </LI>
);

export type CommentTextBlockProps = {
  comment?: CommentData;
  renderers?: MDRenderers;
  withDash?: boolean;
  beforeContent?: JSX.Element;
};

export const CommentTextBlock: React.FC<CommentTextBlockProps> = ({
  comment,
  renderers = mdRenderers,
  withDash,
  beforeContent,
}) => {
  const shortText = comment?.shortText?.trim().length ? (
    <ReactMarkdown renderers={renderers}>{comment.shortText}</ReactMarkdown>
  ) : null;
  const text = comment?.text?.trim().length ? (
    <ReactMarkdown renderers={renderers}>{comment.text}</ReactMarkdown>
  ) : null;

  const example = comment?.tags?.filter(tag => tag.tag === 'example')[0];
  const exampleText = example ? (
    <ReactMarkdown renderers={renderers}>{`__Example:__ ${example.text}`}</ReactMarkdown>
  ) : null;

  const deprecation = comment?.tags?.filter(tag => tag.tag === 'deprecated')[0];
  const deprecationNote = deprecation ? (
    <Quote key="deprecation-note">
      <B>{deprecation.text.trim().length ? deprecation.text : 'Deprecated'}</B>
    </Quote>
  ) : null;

  return (
    <>
      {deprecationNote}
      {beforeContent}
      {withDash && (shortText || text) ? ' - ' : null}
      {shortText}
      {text}
      {exampleText}
    </>
  );
};

export const STYLES_OPTIONAL = css`
  color: ${theme.text.secondary};
  font-size: 90%;
  padding-top: 22px;
`;

export const STYLES_SECONDARY = css`
  color: ${theme.text.secondary};
  font-size: 90%;
  font-weight: 600;
`;
