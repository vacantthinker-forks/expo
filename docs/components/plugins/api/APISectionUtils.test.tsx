import { render } from '@testing-library/react';
import * as React from 'react';

import { resolveTypeName } from './APISectionUtils';

describe('APISectionUtils.resolveTypeName', () => {
  test('void', () => {
    const { container } = render(<>{resolveTypeName({ type: 'intrinsic', name: 'void' })}</>);
    expect(container).toMatchSnapshot();
  });

  test('generic type', () => {
    const { container } = render(<>{resolveTypeName({ type: 'intrinsic', name: 'string' })}</>);
    expect(container).toMatchSnapshot();
  });

  test('custom type', () => {
    const { container } = render(
      <>{resolveTypeName({ type: 'reference', name: 'SpeechSynthesisEvent' })}</>
    );
    expect(container).toMatchSnapshot();
  });

  test('custom type array', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'array',
          elementType: { type: 'reference', name: 'AppleAuthenticationScope' },
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('custom type non-linkable array', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'array',
          elementType: { type: 'reference', name: 'T' },
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('query type', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reference',
          typeArguments: [{ queryType: { type: 'reference', name: 'View' }, type: 'query' }],
          name: 'React.ComponentProps',
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('Promise', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reference',
          typeArguments: [{ type: 'intrinsic', name: 'void' }],
          name: 'Promise',
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('Promise with custom type', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reference',
          typeArguments: [{ type: 'reference', name: 'AppleAuthenticationCredential' }],
          name: 'Promise',
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('Record', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reference',
          typeArguments: [
            { type: 'intrinsic', name: 'string' },
            { type: 'intrinsic', name: 'any' },
          ],
          name: 'Record',
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('Record with union', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reference',
          typeArguments: [
            { type: 'intrinsic', name: 'string' },
            {
              type: 'union',
              types: [
                { type: 'intrinsic', name: 'number' },
                { type: 'intrinsic', name: 'boolean' },
                { type: 'intrinsic', name: 'string' },
              ],
            },
          ],
          name: 'Record',
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('union', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'union',
          types: [
            { type: 'reference', name: 'SpeechEventCallback' },
            { type: 'literal', value: null },
          ],
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('union with array', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'union',
          types: [
            { type: 'array', elementType: { type: 'intrinsic', name: 'number' } },
            { type: 'literal', value: null },
          ],
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('union with custom type and array', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'union',
          types: [
            { type: 'array', elementType: { type: 'reference', name: 'AssetRef' } },
            { type: 'reference', name: 'AssetRef' },
          ],
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('generic type', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reference',
          typeArguments: [{ type: 'reference', name: 'Asset' }],
          name: 'PagedInfo',
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('tuple type', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'tuple',
          elements: [
            { type: 'reference', name: 'SortByKey' },
            { type: 'intrinsic', name: 'boolean' },
          ],
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('generic type in Promise', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reference',
          typeArguments: [
            {
              type: 'reference',
              typeArguments: [{ type: 'reference', name: 'Asset' }],
              name: 'PagedInfo',
            },
          ],
          name: 'Promise',
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('function', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reflection',
          declaration: {
            signatures: [
              {
                type: {
                  type: 'union',
                  types: [
                    { type: 'intrinsic', name: 'void' },
                    {
                      type: 'reference',
                      name: 'SpeechEventCallback',
                    },
                  ],
                },
              },
            ],
          },
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('function with arguments', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reflection',
          declaration: {
            signatures: [
              {
                parameters: [
                  {
                    name: 'error',
                    type: { type: 'reference', name: 'Error' },
                  },
                ],
                type: {
                  type: 'union',
                  types: [
                    { type: 'intrinsic', name: 'void' },
                    { type: 'reference', name: 'SpeechEventCallback' },
                  ],
                },
              },
            ],
          },
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('function with non-linkable custom type', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reflection',
          declaration: {
            signatures: [
              {
                parameters: [
                  {
                    name: 'error',
                    type: { type: 'reference', name: 'Error' },
                  },
                ],
                type: { type: 'intrinsic', name: 'void' },
              },
            ],
          },
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('extended props with multiple omits', () => {
    const { container } = render(
      <>
        {resolveTypeName({
          type: 'reference',
          typeArguments: [
            {
              type: 'reference',
              typeArguments: [
                { type: 'reference', name: 'ViewStyle' },
                {
                  type: 'union',
                  types: [
                    { type: 'literal', value: 'backgroundColor' },
                    {
                      type: 'literal',
                      value: 'borderRadius',
                    },
                  ],
                },
              ],
              name: 'Omit',
            },
          ],
          name: 'StyleProp',
        })}
      </>
    );
    expect(container).toMatchSnapshot();
  });
});
