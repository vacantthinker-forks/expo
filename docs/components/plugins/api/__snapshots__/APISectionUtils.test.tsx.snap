// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APISectionUtils.resolveTypeName Promise 1`] = `
<div>
  Promise
  &lt;
  <span>
    void
  </span>
  &gt;
</div>
`;

exports[`APISectionUtils.resolveTypeName Promise with custom type 1`] = `
<div>
  Promise
  &lt;
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#appleauthenticationcredential"
    >
      AppleAuthenticationCredential
    </a>
  </span>
  &gt;
</div>
`;

exports[`APISectionUtils.resolveTypeName Record 1`] = `
<div>
  Record
  &lt;
  <span>
    string
    , 
  </span>
  <span>
    any
  </span>
  &gt;
</div>
`;

exports[`APISectionUtils.resolveTypeName Record with union 1`] = `
<div>
  Record
  &lt;
  <span>
    string
    , 
  </span>
  <span>
    <span>
      number
       | 
    </span>
    <span>
      boolean
       | 
    </span>
    <span>
      string
    </span>
  </span>
  &gt;
</div>
`;

exports[`APISectionUtils.resolveTypeName custom type 1`] = `
<div>
  <a
    class="css-12zqqiz-STYLES_EXTERNAL_LINK"
    href="/#speechsynthesisevent"
  >
    SpeechSynthesisEvent
  </a>
</div>
`;

exports[`APISectionUtils.resolveTypeName custom type array 1`] = `
<div>
  <a
    class="css-12zqqiz-STYLES_EXTERNAL_LINK"
    href="/#appleauthenticationscope"
  >
    AppleAuthenticationScope
    []
  </a>
</div>
`;

exports[`APISectionUtils.resolveTypeName custom type non-linkable array 1`] = `
<div>
  T[]
</div>
`;

exports[`APISectionUtils.resolveTypeName extended props with multiple omits 1`] = `
<div>
  StyleProp
  &lt;
  <span>
    Omit
    &lt;
    <span>
      ViewStyle
      , 
    </span>
    <span>
      <span>
        'backgroundColor'
         | 
      </span>
      <span>
        'borderRadius'
      </span>
    </span>
    &gt;
  </span>
  &gt;
</div>
`;

exports[`APISectionUtils.resolveTypeName function 1`] = `
<div>
  () =&gt;
   
  <span>
    void
     | 
  </span>
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#speecheventcallback"
    >
      SpeechEventCallback
    </a>
  </span>
</div>
`;

exports[`APISectionUtils.resolveTypeName function with arguments 1`] = `
<div>
  (
  <span>
    error
    : 
    Error
  </span>
  ) 
  =&gt;
   
  <span>
    void
     | 
  </span>
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#speecheventcallback"
    >
      SpeechEventCallback
    </a>
  </span>
</div>
`;

exports[`APISectionUtils.resolveTypeName function with non-linkable custom type 1`] = `
<div>
  (
  <span>
    error
    : 
    Error
  </span>
  ) 
  =&gt;
   
  void
</div>
`;

exports[`APISectionUtils.resolveTypeName generic type 1`] = `
<div>
  string
</div>
`;

exports[`APISectionUtils.resolveTypeName generic type 2`] = `
<div>
  <a
    class="css-12zqqiz-STYLES_EXTERNAL_LINK"
    href="/#pagedinfo"
  >
    PagedInfo
  </a>
  &lt;
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#asset"
    >
      Asset
    </a>
  </span>
  &gt;
</div>
`;

exports[`APISectionUtils.resolveTypeName generic type in Promise 1`] = `
<div>
  Promise
  &lt;
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#pagedinfo"
    >
      PagedInfo
    </a>
    &lt;
    <span>
      <a
        class="css-12zqqiz-STYLES_EXTERNAL_LINK"
        href="/#asset"
      >
        Asset
      </a>
    </span>
    &gt;
  </span>
  &gt;
</div>
`;

exports[`APISectionUtils.resolveTypeName query type 1`] = `
<div>
  React.ComponentProps
  &lt;
  <span>
    View
  </span>
  &gt;
</div>
`;

exports[`APISectionUtils.resolveTypeName tuple type 1`] = `
<div>
  [
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#sortbykey"
    >
      SortByKey
    </a>
    , 
  </span>
  <span>
    boolean
  </span>
  ]
</div>
`;

exports[`APISectionUtils.resolveTypeName union 1`] = `
<div>
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#speecheventcallback"
    >
      SpeechEventCallback
    </a>
     | 
  </span>
  <span>
    null
  </span>
</div>
`;

exports[`APISectionUtils.resolveTypeName union with array 1`] = `
<div>
  <span>
    number[]
     | 
  </span>
  <span>
    null
  </span>
</div>
`;

exports[`APISectionUtils.resolveTypeName union with custom type and array 1`] = `
<div>
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#assetref"
    >
      AssetRef
      []
    </a>
     | 
  </span>
  <span>
    <a
      class="css-12zqqiz-STYLES_EXTERNAL_LINK"
      href="/#assetref"
    >
      AssetRef
    </a>
  </span>
</div>
`;

exports[`APISectionUtils.resolveTypeName void 1`] = `
<div>
  void
</div>
`;
